### PR TITLE
upload based on manifest diff, not snapshot

### PR DESCRIPTION
**Pull request by memsharded**
This could be a fix for:
- https://github.com/conan-io/conan/issues/3067
- https://github.com/conan-io/conan/issues/3446
- https://github.com/conan-io/conan/issues/3473

This is the implementation for API V1.

### CODE REVIEW

File: conans/client/cmd/uploader.py [hunk 1]
```diff
@@ -84,7 +83,7 @@ def _upload(self, conan_file, conan_ref, force, packages_ids, retry, retry_wait,
         else:  # Or use the default otherwise
             upload_remote = self._registry.default_remote
 
-        if not force:
+        if no_overwrite != "force":
```
Comment (danimtb) [inline conans/client/cmd/uploader.py]: This comparison is a bit confusing. I know doing it this way there is one parameter less but I think the old code reads better IMO

Comment (memsharded) [reply conans/client/cmd/uploader.py]: Yeah, I know, but the alternative is passing around 2 mutually exclusive arguments, which is bad interface. I thought changing both for a ``policy``, but wanted to keep changes to a minimum.

File: conans/client/rest/rest_client_v1.py [hunk 2]
```diff
@@ -203,31 +203,31 @@ def _get_package_urls(self, package_reference):
...
             if no_overwrite in ("all", "recipe"):
                 raise ConanException("Local recipe is different from the remote recipe. "
                                      "Forbidden overwrite")
+
+        local_snapshot = {filename: None for filename, abs_path in the_files.items()}
         files_to_upload = {filename.replace("\\", "/"): the_files[filename]
-                           for filename in new + modified}
+                           for filename in local_snapshot}
+
+        deleted = set(remote_snapshot).difference(local_snapshot)
```
Comment (jgsogo) [inline conans/client/rest/rest_client_v1.py]: `remote_snapshot` keys have backward slashes?

File: conans/client/rest/rest_client_v1.py [hunk 3]
```diff
@@ -203,31 +203,31 @@ def _get_package_urls(self, package_reference):
...
-            return False, conan_reference
+            if remote_manifest == local_manifest:
+                return False, conan_reference
 
-        if no_overwrite and remote_snapshot:
             if no_overwrite in ("all", "recipe"):
                 raise ConanException("Local recipe is different from the remote recipe. "
                                      "Forbidden overwrite")
+
+        local_snapshot = {filename: None for filename, abs_path in the_files.items()}
```
Comment (jgsogo) [inline conans/client/rest/rest_client_v1.py]: `local_snapshot = set(the_files.keys())` (or leave it as an iterable)

File: conans/client/rest/rest_client_v1.py [hunk 4]
```diff
@@ -251,19 +251,20 @@ def upload_package(self, package_reference, the_files, retry, retry_wait, no_ove
...
+            if remote_manifest == local_manifest:
+                return False
 
-        if no_overwrite and remote_snapshot:
             if no_overwrite == "all":
                 raise ConanException("Local package is different from the remote package. "
                                      "Forbidden overwrite")
 
-        files_to_upload = {filename: the_files[filename] for filename in new + modified}
+        local_snapshot = {filename: None for filename, abs_path in the_files.items()}
```
Comment (jgsogo) [inline conans/client/rest/rest_client_v1.py]: `local_snapshot = set(the_files.keys())` (or leave it as an iterable, same as above)

File: conans/client/rest/rest_client_v1.py [hunk 5]
```diff
@@ -251,19 +251,20 @@ def upload_package(self, package_reference, the_files, retry, retry_wait, no_ove
...
+        files_to_upload = {filename: the_files[filename] for filename in local_snapshot}
```
Comment (jgsogo) [inline conans/client/rest/rest_client_v1.py]: `files_to_upload` is not just the same dictionary as `the_files`?

File: conans/client/rest/rest_client_v1.py [hunk 6]
```diff
@@ -203,31 +203,31 @@ def _get_package_urls(self, package_reference):
...
 
-        if no_overwrite and remote_snapshot:
             if no_overwrite in ("all", "recipe"):
                 raise ConanException("Local recipe is different from the remote recipe. "
                                      "Forbidden overwrite")
+
+        local_snapshot = {filename: None for filename, abs_path in the_files.items()}
         files_to_upload = {filename.replace("\\", "/"): the_files[filename]
-                           for filename in new + modified}
+                           for filename in local_snapshot}
```
Comment (jgsogo) [inline conans/client/rest/rest_client_v1.py]: `files_to_upload = {filename.replace("\\", "/"): abs_path for filename, abs_path in the_files.items()}`

And I'm also wondering if this replacement is needed at all

File: conans/client/cmd/uploader.py (hunk 1 above)
Comment (lasote) [reply conans/client/cmd/uploader.py]: introduce the policy please, I really hate that kind of arguments

File: conans/client/rest/rest_client_v1.py (hunk 2 above)
Comment (memsharded) [reply conans/client/rest/rest_client_v1.py]: If they are there, better leave them, there might be some reason.

File: conans/client/rest/rest_client_v1.py (hunk 6 above)
Comment (memsharded) [reply conans/client/rest/rest_client_v1.py]: Just, it is a bit ugly, but I prefer to leave it by now, too many changes.

File: conans/client/command.py [hunk 7]
```diff
@@ -1007,14 +1009,30 @@ def upload(self, *args):
         cwd = os.getcwd()
         info = None
 
+        if args.force and args.no_overwrite:
+            raise ConanException("'no_overwrite' argument cannot be used together with 'force'")
```
Comment (danimtb) [inline conans/client/command.py]: `no_overwrite` was the notation used to raise from conan API. I would indicate here --no-overwrite to have a more message. Same for `skip_upload` and `force` -- > `--skip-upload` & `--force`

Comment (memsharded) [reply conans/client/command.py]: Done

**Pull request merged**